src/DotNetWorker.Grpc/Handlers/InvocationHandler.cs

```diff
@@ -0,0 +1,154 @@
+using System.Threading.Tasks;
+using Azure.Core.Serialization;
+using Microsoft.Azure.Functions.Worker.Context.Features;
+using Microsoft.Azure.Functions.Worker.Grpc;
+using Microsoft.Azure.Functions.Worker.Grpc.Features;
+using Microsoft.Azure.Functions.Worker.Grpc.Messages;
+using Microsoft.Azure.Functions.Worker.Logging;
+using Microsoft.Azure.Functions.Worker.OutputBindings;
+using Microsoft.Azure.Functions.Worker.Rpc;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Functions.Worker.Handlers
+{
+    internal class InvocationHandler : IInvocationHandler
+    {
+        private readonly IFunctionsApplication _application;
+        private readonly IInvocationFeaturesFactory _invocationFeaturesFactory;
+        private readonly IOutputBindingsInfoProvider _outputBindingsInfoProvider;
+        private readonly IInputConversionFeatureProvider _inputConversionFeatureProvider;
+        private readonly ObjectSerializer _serializer;
+        private readonly ILogger _logger;
+
+        private ConcurrentDictionary<string, CancellationTokenSource> _inflightInvocations => new();
+
+        public InvocationHandler(
+            IFunctionsApplication application,
+            IInvocationFeaturesFactory invocationFeaturesFactory,
+            ObjectSerializer serializer,
+            IOutputBindingsInfoProvider outputBindingsInfoProvider,
+            IInputConversionFeatureProvider inputConversionFeatureProvider,
+            ILoggerProvider loggerProvider)
+        {
+            _application = application ?? throw new ArgumentNullException(nameof(application));
+            _invocationFeaturesFactory = invocationFeaturesFactory ?? throw new ArgumentNullException(nameof(invocationFeaturesFactory));
+            _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
+            _outputBindingsInfoProvider = outputBindingsInfoProvider ?? throw new ArgumentNullException(nameof(outputBindingsInfoProvider));
+            _inputConversionFeatureProvider = inputConversionFeatureProvider ?? throw new ArgumentNullException(nameof(inputConversionFeatureProvider));
+            _logger = loggerProvider.CreateLogger(nameof(InvocationHandler));
+        }
+
+        public async Task<InvocationResponse> InvokeAsync(InvocationRequest request)
+        {
+            FunctionContext? context = null;
+            InvocationResponse response = new()
+            {
+                InvocationId = request.InvocationId,
+                Result       = new StatusResult()
+            };
+
+            try
+            {
+                var invocation = new GrpcFunctionInvocation(request);
+                var cancellationTokenSource = new CancellationTokenSource();
+
+                if (!_inflightInvocations.TryAdd(request.InvocationId, cancellationTokenSource))
+                {
+                    // Dispose the CancellationTokenSource if we cannot keep track of it
+                    cancellationTokenSource?.Dispose();
```

Comment:
Null operator `?.` is not necessary from what I can tell - you have new'ed it up above so it wont be null.